### PR TITLE
file service system file delete and undelete

### LINKED ISSUE #6149: systemDelete - preHandle

### PR DESCRIPTION
**Pull request by povolev15**
Implemented System delete and undelete pre handler, handler and tests

**Related issue(s)**:

Fixes #6149 #6150

### CODE REVIEW

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java [hunk 1]
```diff
@@ -101,23 +105,22 @@ public void handle(
...
 
-        /* Copy all the fields from existing topic and change deleted flag */
-        final var fileBuilder = new File.Builder()
-                .fileNumber(file.fileNumber())
-                .expirationTime(file.expirationTime())
-                .keys(file.keys())
-                .contents(file.contents())
-                .memo(file.memo())
-                .deleted(true);
+        /* Copy all the fields from existing file and change deleted flag */
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java:108]: Slightly misleading.  This code also removes the file contents (sets the contents to null).

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java [hunk 2]
```diff
@@ -35,21 +51,72 @@ public FileSystemDeleteHandler() {
...
+     *
+     * <p>Determines signatures needed for deleting a system file
+     *
+     * @param context the {@link PreHandleContext} which collects all information that will be
+     *     passed to {@code handle()}
+     * @throws NullPointerException if any of the arguments are {@code null}
+     */
     @Override
-    public void preHandle(@NonNull final PreHandleContext context) {
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:64]: This method does need to check that the entity ID is less than the configured max system entity ID, and fail otherwise.  The SystemFileDelete transaction is only valid for files with entity ID values in the correct range (0 < ID < MAX_SYSTEM_ENTITY).

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 3]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+    /**
+     * The function validates the file id and returns the file metadata for system files.
+     *
+     * @param handleContext the handle context for the transaction.
+     * @param fileStore the file store to fetch the metadata of specified file id
+     * @param fileId the file id to validate and to fetch the metadata
+     * @param canBeDeleted flag if the file can be deleted or not
+     * @return the file metadata of specific system file id
+     */
+    public static @NonNull File verifyFileSystem(
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:119]: This might be clearer as `verifyFileIsSystem` or `verifySystemFile`, as `verifyFileSystem` sounds like it is verifying the _filesystem_, rather than verifying that a particular file is a "system file".

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 4]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+            @NonNull final HandleContext handleContext,
+            @NonNull final WritableFileStoreImpl fileStore,
+            @NonNull final FileID fileId,
+            final boolean canBeDeleted) {
+        requireNonNull(handleContext);
+        requireNonNull(fileStore);
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:125]: Probably could elide these two lines (125,126), as the very next line will throw the NPE and include better details.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 5]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+        var optionalFile = fileStore.get(fileId.fileNum());
+
+        if (optionalFile.isEmpty()) {
+            throw new HandleException(INVALID_FILE_ID);
+        }
+
+        // where to get the max special file number
+        final var fileServiceConfig =
+                handleContext.getConfiguration().getConfigData(FilesConfig.class);
+        // TODO: change to the property file config value
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:137]: This TODO needs to be completed before merging.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 6]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+        if (fileId.fileNum() > 750) {
+            throw new HandleException(INVALID_FILE_ID);
+        }
+
+        final var file = optionalFile.get();
+
+        if (!file.hasKeys() || file.keys().keys().isEmpty()) {
+            // @todo('protobuf change needed') change to immutable file response code
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:145]: Very odd that we don't have a response code for this.  Should be brought up in the daily call.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 7]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+            for (Key key : listKeys.keys()) context.requireKey(key);
+    }
+
+    /**
+     * The function validates the file id and returns the file metadata for system files.
+     *
+     * @param handleContext the handle context for the transaction.
+     * @param fileStore the file store to fetch the metadata of specified file id
+     * @param fileId the file id to validate and to fetch the metadata
+     * @param canBeDeleted flag if the file can be deleted or not
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:116]: `flag indicating that the file may be deleted, and the call should not fail if it is.`

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        final var existingFile = writableStore.get(fileEntityNum.longValue());
+        assertTrue(existingFile.isPresent());
+        assertFalse(existingFile.get().deleted());
+
+        final var msg =
+                assertThrows(
+                        HandleException.class,
+                        () -> subject.handle(handleContext, txn, writableStore));
+        assertEquals(INVALID_FILE_ID, msg.getStatus());
+    }
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java:155]: Should also assert here that the file is still no deleted.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 6 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:145]: I know but I checked it out I may check with @jasperpotts about it.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 9]
```diff
@@ -77,19 +86,85 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+        if (!canBeDeleted && file.deleted()) {
+            throw new HandleException(FILE_DELETED);
+        }
+
+        return file;
+    }
+
+    /**
+     * The function validates the file id and returns the file metadata for system files.
+     *
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:158]: done and added some more thank you

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 7 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:116]: rewrite it thanks

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java:155]: if the file id is greater than 750 it doesn't matter. I don't check the rest.
I can create a separate test that validates for file deleted.

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java:155]: line 149 is doing exactly it `assertFalse(existingFile.get().deleted());`

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 4 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:125]: ok fixed thx

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:119]: sure thank you.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java (hunk 2 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:64]: I fixed it but I need a property for the MAX_SYSTEM_ENTITY.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 10]
```diff
@@ -81,6 +84,11 @@ public static void validateContent(
 
         final var fileMeta = fileStore.getFileMetadata(fileId);
         mustExist(fileMeta, INVALID_FILE_ID);
+
+        if (isSpecialFile && fileId.fileNum() > 750) {
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:88]: Make certain to replace the constant here with config (You'll need to pass in context or the value read from config) before merging.  This value has changed in the past and likely will change again in the future.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java (hunk 2 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:64]: Understood, I suspect you can add that in the appropriate config record (`LedgerConfig`?), but check with @hendrikebbers to be sure.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:146]: I fixed it and added it to the Ledger config. @hendrikebbers please advice if you think it is supposed to be somewhere else.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java:152]: sure I will add it

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 10 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java]: I changed it to read it from the Ledger config file.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java (hunk 1 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java:105]: Still not accurate, as it does not copy all fields, it deletes the contents (irretrievably, in fact).

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 11]
```diff
@@ -58,28 +64,42 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
         if (fileId == null) {
             throw new PreCheckException(INVALID_FILE_ID);
         }
 
         final var fileMeta = fileStore.getFileMetadata(fileId);
         mustExist(fileMeta, INVALID_FILE_ID);
+
+        final var ledgerConfig = context.getConfiguration().getConfigData(LedgerConfig.class);
+        if (fileId.fileNum() > ledgerConfig.numReservedSystemEntities()) {
+            throw new HandleException(INVALID_FILE_ID);
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:90]: Is HandleException correct, or should it be PreCheckException?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 12]
```diff
@@ -58,28 +64,42 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
 
     /**
      * The function validation the file id and returns the file metadata.
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:66]: "Validates that the fileId is non-null, not a reserved system Id, and matches a file in the store."

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java [hunk 13]
```diff
@@ -209,7 +210,7 @@ void handleWorksAsExpected() {
         assertEquals(1_234_567L, actualFile.expirationTime());
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:210]: We've decided to only use AssertJ for assertions. All of these assertions should be changed to use AssertJ.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java [hunk 14]
```diff
@@ -209,7 +210,7 @@ void handleWorksAsExpected() {
...
         assertEquals(contentsBytes, actualFile.contents());
         assertEquals(fileEntityNum.longValue(), actualFile.fileNumber());
-        assertEquals(false, actualFile.deleted());
+        assertFalse(actualFile.deleted());
         assertEquals(1_234L, recordBuilder.getCreatedFile());
         assertTrue(fileStore.get(1234L).isPresent());
     }
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:216]: The TODO just below this line needs more information. Ideally each TODO should cite an issue number that more fully describes what needs to be done.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java (hunk 8 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileSystemDeleteHandlerTest.java:157]: done

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemUndeleteHandler.java [hunk 15]
```diff
@@ -35,21 +47,68 @@ public FileSystemUndeleteHandler() {
...
     }
 
+    /**
+     * This method is called during the pre-handle workflow.
+     *
+     * <p>Determines signatures needed for undelete system file
+     *
+     * @param context the {@link PreHandleContext} which collects all information that will be
+     *     passed to {@code handle()}
+     * @throws NullPointerException if any of the arguments are {@code null}
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemUndeleteHandler.java:57]: done thanks

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 12 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:66]: we are not doing so as @jsync-swirlds mentioned in multiple PRS is NPM is happening in the first 3 rows there is no matter to apply requiredNonNull.
It also validates if it is a file system 2 lines below.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java (hunk 13 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:210]: if that's the case I think we need to change the all-unit test in all services. most of the services use asserts from junit

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java (hunk 14 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:216]: removed it

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java (hunk 13 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:210]: can I open a separate PR for it later on depends on the priority.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java [hunk 16]
```diff
@@ -35,21 +48,70 @@ public FileSystemDeleteHandler() {
...
+                .seconds();
+        if (newExpiry <= handleContext.consensusNow().getEpochSecond()) {
+            fileStore.removeFile(fileId.fileNum());
+        } else {
+            /* Get all the fields from existing file and change deleted flag */
+            final var fileBuilder = new File.Builder()
+                    .fileNumber(file.fileNumber())
+                    .expirationTime(newExpiry)
+                    .keys(file.keys())
+                    .contents(file.contents())
```
Comment (hendrikebbers) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:102]: In FileDeleteHandler you call `contents(null)` instead of `contents(file.contents())`. I would assume to have the same call in both

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 17]
```diff
@@ -92,4 +112,73 @@ public static void validateAndAddRequiredKeys(
...
+     * @param handleContext the handle context for the transaction.
+     * @param fileStore the file store to fetch the metadata of specified file id
+     * @param fileId the file id to validate and to fetch the metadata
+     * @param canBeDeleted flag indicating that the file may be deleted, and the call should not
+     *     fail if it is.
+     * @return the file metadata of specific system file id
+     */
+    public static @NonNull File verifySystemFile(
+            @NonNull final HandleContext handleContext,
+            @NonNull final WritableFileStoreImpl fileStore,
```
Comment (hendrikebbers) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:133]: Do we really need the Impl?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 18]
```diff
@@ -92,4 +112,73 @@ public static void validateAndAddRequiredKeys(
...
+     *
+     * @param handleContext the handle context for the transaction.
+     * @param fileStore the file store to fetch the metadata of specified file id
+     * @param fileId the file id to validate and to fetch the metadata
+     * @param canBeDeleted flag indicating that the file may be deleted, and the call should not
+     *     fail if it is.
+     * @return the file metadata of specific system file id
+     */
+    public static @NonNull File verifySystemFile(
+            @NonNull final HandleContext handleContext,
```
Comment (hendrikebbers) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:132]: Since this is only used to get the config it would be better to just pass the config record to the method

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 19]
```diff
@@ -92,4 +112,73 @@ public static void validateAndAddRequiredKeys(
...
+     * file is empty this call will fail with {@link ResponseCodeEnum#INVALID_FILE_ID}. If the
+     * special file id is greater than special file config , this call will fail with {@link
+     * ResponseCodeEnum#INVALID_FILE_ID}. If the file has no keys or empty list, this call will fail
+     * with {@link ResponseCodeEnum#UNAUTHORIZED}. If the file is already deleted, this call will
+     * fail with {@link ResponseCodeEnum#FILE_DELETED}.
+     *
+     * @param handleContext the handle context for the transaction.
+     * @param fileStore the file store to fetch the metadata of specified file id
+     * @param fileId the file id to validate and to fetch the metadata
+     * @param canBeDeleted flag indicating that the file may be deleted, and the call should not
```
Comment (hendrikebbers) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:127]: I do not understand that parameter

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendHandlerTest.java [hunk 20]
```diff
@@ -19,9 +19,12 @@
 import static com.hedera.hapi.node.base.ResponseCodeEnum.FILE_DELETED;
 import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_FILE_ID;
 import static com.hedera.node.app.spi.validation.ExpiryMeta.NA;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
```
Comment (hendrikebbers) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendHandlerTest.java:22]: the only change in that file are new imports?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableFileStoreImpl.java [hunk 21]
```diff
@@ -101,9 +104,20 @@ public long sizeOfState() {
...
         return filesState.modifiedKeys();
     }
+
+    /**
+     * remove the file from the state.
+     *
+     * @param fileNum - the number of the file to be removed from state.
+     */
+    public void removeFile(final long fileNum) {
+        requireNonNull(fileNum);
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableFileStoreImpl.java:120]: you are right it I just was adding it to all on the previous pr thanks

[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/WritableFileStoreImpl.java:120]: removed all 3 intstances

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java [hunk 22]
```diff
@@ -35,21 +48,70 @@ public FileSystemDeleteHandler() {
...
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(
+            @NonNull final HandleContext handleContext,
+            @NonNull final SystemDeleteTransactionBody systemDeleteTransactionBody,
+            @NonNull final WritableFileStoreImpl fileStore) {
+        requireNonNull(handleContext);
+        requireNonNull(systemDeleteTransactionBody);
+        requireNonNull(fileStore);
+
+        final var fileId = systemDeleteTransactionBody.fileIDOrElse(FileID.DEFAULT);
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:87]: that what I have done also in file delete and also consensus service is doing. it all will be 0 and it will fail during the validation. the other option can be that we will throw an exception for not trx in body

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java [hunk 23]
```diff
@@ -35,21 +48,70 @@ public FileSystemDeleteHandler() {
...
+
+        final File file = verifySystemFile(handleContext, fileStore, fileId);
+
+        final var newExpiry = systemDeleteTransactionBody
+                .expirationTimeOrElse(new TimestampSeconds(file.expirationTime()))
+                .seconds();
+        if (newExpiry <= handleContext.consensusNow().getEpochSecond()) {
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:94]: sure no problem I will add it but a quick explanation. if the expiration time of the transaction exist we take if otherwise from the file store and check if the current time is greater than the expiry if yes we delete the file if not we just change the deleted flag to false

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java (hunk 16 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileSystemDeleteHandler.java:102]: no because file system can be undeleted while regular file cannot that why it is different

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java (hunk 1 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java:105]: fixed it.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java (hunk 13 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileCreateHandlerTest.java:210]: no problem but can we do it as part of a separate PR.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendHandlerTest.java (hunk 20 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileAppendHandlerTest.java:22]: removed

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 19 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:127]: I changed a little bit the description but it is needed for the file system to undelete the file because the flag is deleted but we need to undelete it and when it is verifying the file it will crash it.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 24]
```diff
@@ -92,4 +112,73 @@ public static void validateAndAddRequiredKeys(
...
+
+        var optionalFile = fileStore.get(fileId.fileNum());
+
+        if (optionalFile.isEmpty()) {
+            throw new HandleException(INVALID_FILE_ID);
+        }
+
+        // where to get the max special file number
+        final var ledgerConfig = handleContext.getConfiguration().getConfigData(LedgerConfig.class);
+        if (fileId.fileNum() > ledgerConfig.numReservedSystemEntities()) {
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:146]: done thanks

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 18 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:132]: sure I will change it

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:132]: done

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 17 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:133]: we need it in order to get the file from the store

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/LedgerConfig.java [hunk 25]
```diff
@@ -36,4 +36,5 @@ public record LedgerConfig(
         @ConfigProperty(value = "tokenTransfers.maxLen", defaultValue = "10") int tokenTransfersMaxLen,
         @ConfigProperty(value = "nftTransfers.maxLen", defaultValue = "10") int nftTransfersMaxLen,
         @ConfigProperty(value = "records.maxQueryableByAccount", defaultValue = "180") int recordsMaxQueryableByAccount,
-        @ConfigProperty(value = "schedule.txExpiryTimeSecs", defaultValue = "1800") int scheduleTxExpiryTimeSecs) {}
+        @ConfigProperty(value = "schedule.txExpiryTimeSecs", defaultValue = "1800") int scheduleTxExpiryTimeSecs,
+        @ConfigProperty(defaultValue = "750") long numReservedSystemEntities) {}
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/LedgerConfig.java:40]: I am not against it. I just wanted to get your opinion I am ok with any place where to put it and to call it the way you want. I called it the way it is currently in [ledgerConfig](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/config/HederaNumbers.java)

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 26]
```diff
@@ -57,29 +63,43 @@ public static void validateContent(@NonNull byte[] content, @NonNull FilesConfig
...
+        requireNonNull(context);
 
         if (fileId == null) {
             throw new PreCheckException(INVALID_FILE_ID);
         }
 
         final var fileMeta = fileStore.getFileMetadata(fileId);
         mustExist(fileMeta, INVALID_FILE_ID);
+
+        final var ledgerConfig = context.getConfiguration().getConfigData(LedgerConfig.class);
```
[the remark this answers is not in the document]
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:88]: done thanks

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/LedgerConfig.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/LedgerConfig.java:40]: The property is currently in three places, and it's not at all clear which control.  ledger.numReservedSystemEntities seems to be the correct property, but there is another in some obscure documentation, and it appears the "default" constant is actually used in all cases currently.
It probably makes the most sense to add it here, as it's similar in character to other ledger configuration values.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileTestUtils.java [hunk 27]
```diff
@@ -58,8 +59,8 @@ private FileTestUtils() {
     static Key mockPayerLookup(Key key, AccountID accountId, ReadableAccountStore accountStore)
             throws PreCheckException {
         final var account = mock(Account.class);
-        given(accountStore.getAccountById(accountId)).willReturn(account);
-        given(account.key()).willReturn(key);
+        lenient().when(accountStore.getAccountById(accountId)).thenReturn(account);
```
Comment (kimbor) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileTestUtils.java:62]: what is the reason for this change?

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileTestUtils.java:62]: the unit test that using this mockPayerLookup is using leginnt I just changed the syntax but the logic is the same
`        lenient().when(accountStore.getAccountById(accountId)).thenReturn(account);
        lenient().when(account.key()).thenReturn(key);`

Comment (kimbor) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileTestUtils.java:62]: Ok

**Review by kimbor — DISMISSED** [review]

**Review by kimbor — APPROVED** [review]

**Review by jsync-swirlds — APPROVED** [review]

**Review by jeromy-cannon — APPROVED** [review]
